File: go/reverse_words_in_a_string_iii.go
package leetcode

// Time complexity: O(n) where n is the length of the input string
// Space complexity: O(n) for the additional runes array
// Note that the algorithm is in-place but since Go strings are
// immutable we have to convert them to an array
func reverseWords(s string) string {
	reverse := func(runes []rune, l, r int) {
		for i := 0; i < (r-l+1)/2; i++ {
			runes[l+i], runes[r-i] = runes[r-i], runes[l+i]
		}
	}

	sRunes := []rune(s)
	l := -1
	for i := range sRunes {
		if i == len(sRunes)-1 || sRunes[i+1] == ' ' {
			if l == -1 {
				l = i
			}
			reverse(sRunes, l, i)
			l = -1
		} else if sRunes[i] != ' ' && l == -1 {
			l = i
		}
	}
	return string(sRunes)
}

File: python/reverse-words-in-a-string-iii.py
class Solution:
    # Time complexity: O(n) where n is the length of the input string
    # Space complexity: O(1)
    def reverseWords(self, s: str) -> str:
        def reverse(l, r, arr):
            N = len(arr)
            for i in range((r-l+1) // 2):
                arr[l+i], arr[r-i] = arr[r-i], arr[l+i]
        
        s = list(s)
        l = -1
        for r in range(len(s)):
            if r == len(s)-1 or s[r+1] == " ":
                if l == -1:
                    l = r
                reverse(l, r, s)
                l = -1
            elif s[r] != " " and l == -1:
                l = r
        return "".join(s)
        
    # Time complexity: O(n) where n is the length of the input string
    # Space complexity: O(n)
    def reverseWords2(self, s: str) -> str:
        return " ".join("".join(list(reversed(word))) for word in s.split(" "))